sled-agent/src/dump_setup.rs

```diff
@@ -211,45 +308,179 @@ enum ZfsGetError {
+        let capacity = used + available;
+        let below = (used * 100) / capacity < percent;
+        Ok((below, used))
+    }
 
-fn zfs_get_integer(
-    mountpoint_or_name: impl AsRef<str>,
-    property: &str,
-) -> Result<u64, ZfsGetError> {
-    zfs_get_prop(mountpoint_or_name, property)?.parse().map_err(Into::into)
+    fn mountpoint(
+        &self,
+        zpool: &ZpoolName,
+        mountpoint: &'static str,
+    ) -> Utf8PathBuf;
 }
 
-fn zfs_get_prop(
-    mountpoint_or_name: impl AsRef<str> + Sized,
-    property: &str,
-) -> Result<String, ZfsGetError> {
-    let mountpoint = mountpoint_or_name.as_ref();
-    let mut cmd = std::process::Command::new(illumos_utils::zfs::ZFS);
-    cmd.arg("get").arg("-Hpo").arg("value");
-    cmd.arg(property);
-    cmd.arg(mountpoint);
-    let output = cmd.output()?;
-    Ok(String::from_utf8(output.stdout)?.trim().to_string())
+trait ZoneInvoker {
+    fn get_zones(&self) -> Result<Vec<Zone>, ArchiveLogsError>;
 }
 
-const DATASET_USAGE_PERCENT_CHOICE: u64 = 70;
-const DATASET_USAGE_PERCENT_CLEANUP: u64 = 80;
+struct RealCoreDumpAdm {}
+struct RealZfs {}
+struct RealZone {}
+
+impl CoreDumpAdmInvoker for RealCoreDumpAdm {
+    fn coreadm(&self, core_dir: &Utf8PathBuf) -> Result<(), ExecutionError> {
+        let mut cmd = CoreAdm::new();
+
+        // disable per-process core patterns
+        cmd.disable(CoreFileOption::Process);
+        cmd.disable(CoreFileOption::ProcSetid);
+
+        // use the global core pattern
+        cmd.enable(CoreFileOption::Global);
+        cmd.enable(CoreFileOption::GlobalSetid);
+
+        // set the global pattern to place all cores into core_dir,
+        // with filenames of "core.[zone-name].[exe-filename].[pid].[time]"
+        cmd.global_pattern(core_dir.join("core.%z.%f.%p.%t"));
+
+        // also collect DWARF data from the exe and its library deps
+        cmd.global_contents("default+debug");
+
+        cmd.execute()
+    }
+
+    // Invokes `dumpadm(8)` to configure the kernel to dump core into the given
+    // `dump_slice` block device in the event of a panic. If a core is already
+    // present in that block device, and a `savecore_dir` is provided, this
+    // function also invokes `savecore(8)` to save it into that directory.
+    // On success, returns Ok(Some(stdout)) if `savecore(8)` was invoked, or
+    // Ok(None) if it wasn't.
+    fn dumpadm(
+        &self,
+        dump_slice: &Utf8PathBuf,
+        savecore_dir: Option<&Utf8PathBuf>,
+    ) -> Result<Option<OsString>, ExecutionError> {
+        let savecore_dir_cloned = if let Some(dir) = savecore_dir.cloned() {
+            dir
+        } else {
+            // if we don't have a savecore destination yet, still create and use
+            // a tmpfs path (rather than the default location under /var/crash,
+            // which is in the ramdisk pool), because dumpadm refuses to do what
+            // we ask otherwise.
```

Comment:
`/tmp` appears to be backed by swap -- I think that using `/var/crash` also has problems (writing a crash dump of the kernel to a ramdisk is not useful) but writing a kernel crash dump to swap also seems like it's basically throwing it away.

Is using this `/tmp` path better than not calling savecore at all?

Reply:
in the case where the /tmp path is used as a placeholder for the invocation to `dumpadm`, we don't actually invoke `savecore`. this is just there to give `dumpadm` a safer placeholder -- we need it to tell the kernel to direct its core to a specific raw partition should it crash, but it also writes the config file that `savecore` reads.  we only invoke `savecore` later in this `if savecore_dir.is_some()` -- that is, if we have a non-placeholder destination for it.